clang/lib/3C/3C.cpp

```diff
@@ -497,48 +521,61 @@ bool _3CInterface::solveConstraints() {
   return true;
 }
 
-bool _3CInterface::writeConvertedFileToDisk(const std::string &FilePath) {
-  std::lock_guard<std::mutex> Lock(InterfaceMutex);
-  bool RetVal = false;
-  if (std::find(SourceFiles.begin(), SourceFiles.end(), FilePath) !=
-      SourceFiles.end()) {
-    RetVal = true;
-    std::vector<std::string> SourceFiles;
-    SourceFiles.clear();
-    SourceFiles.push_back(FilePath);
-    // Don't use global tool. Create a new tool for give single file.
-    ClangTool Tool(*CurrCompDB, SourceFiles);
-    Tool.appendArgumentsAdjuster(getIgnoreCheckedPointerAdjuster());
-    std::unique_ptr<ToolAction> RewriteTool =
-        newFrontendActionFactoryA<RewriteAction<RewriteConsumer, ProgramInfo>>(
-            GlobalProgramInfo, VerifyDiagnosticOutput);
-
-    if (RewriteTool) {
-      int ToolExitCode = Tool.run(RewriteTool.get());
-      if (ToolExitCode != 0)
-        RetVal = false;
-    }
-  }
-  GlobalProgramInfo.getPerfStats().endTotalTime();
-  GlobalProgramInfo.getPerfStats().startTotalTime();
-  return RetVal;
-}
+//bool _3CInterface::writeConvertedFileToDisk(const std::string &FilePath) {
+//  std::lock_guard<std::mutex> Lock(InterfaceMutex);
+//  bool RetVal = false;
+//  if (std::find(SourceFiles.begin(), SourceFiles.end(), FilePath) !=
+//      SourceFiles.end()) {
+//    RetVal = true;
+//    std::vector<std::string> SourceFiles;
+//    SourceFiles.clear();
+//    SourceFiles.push_back(FilePath);
+//    // Don't use global tool. Create a new tool for give single file.
+//    ClangTool Tool(*CurrCompDB, SourceFiles);
+//    Tool.appendArgumentsAdjuster(getIgnoreCheckedPointerAdjuster());
+//
+//    RewriteConsumer RC = RewriteConsumer(GlobalProgramInfo);
+//    for (auto &TU : ASTs) {
+//      RC.HandleTranslationUnit(TU->getASTContext());
+//    }
+//
+//
+////    std::unique_ptr<ToolAction> RewriteTool =
+////        newFrontendActionFactoryA<RewriteAction<RewriteConsumer, ProgramInfo>>(
+////            GlobalProgramInfo, VerifyDiagnosticOutput);
+////
+////    if (RewriteTool) {
+////      int ToolExitCode = Tool.run(RewriteTool.get());
+////      if (ToolExitCode != 0)
+////        RetVal = false;
+////    }
+//
+//  }
+//  GlobalProgramInfo.getPerfStats().endTotalTime();
+//  GlobalProgramInfo.getPerfStats().startTotalTime();
+//  return RetVal;
+//}
 
 bool _3CInterface::writeAllConvertedFilesToDisk() {
   std::lock_guard<std::mutex> Lock(InterfaceMutex);
 
-  ClangTool &Tool = getGlobalClangTool();
+//  ClangTool &Tool = getGlobalClangTool();
+
+  RewriteConsumer RC = RewriteConsumer(GlobalProgramInfo);
+  for (auto &TU : ASTs) {
+    RC.HandleTranslationUnit(TU->getASTContext());
+  }
```

Comment:
After rewriting happens, do you need to delete/free the ASTs? Maybe add another method to free them, to complement the one I'm suggesting you add, to build them?

Reply:
From what I could tell the destructors handled everything. That might be different if we had to save/load for memory, or maintain the diagnostics for verification. If you can think of anything else, I'll check it out and add it.